File: command-center/lib/sceneRetention.ts
/**
 * Scene-level retention — which scene the audience left in (roadmap Q8, PR 5.1).
 *
 * The TypeScript twin of `modules/scene_retention.py`; both run the same cases
 * in `samples/scene_retention_cases.json`, so a divergence fails on both sides.
 *
 * The Video IR (migration 0013) gives each scene its REAL start/end on the
 * narration audio, and the rendered video is exactly as long as that audio, so
 * a YouTube retention point at `elapsed_ratio` sits at `elapsed_ratio *
 * duration` seconds. Per scene: retention at its start and end (linearly
 * interpolated between measured points), the drop (start − end, in
 * share-of-audience points), the drop per minute, and a rank — ranked on the
 * RATE, because a long scene loses more viewers simply by being long.
 *
 * Unknown stays unknown: a scene without real times, a video without a known
 * duration, or a curve with fewer than MIN_POINTS measured points gives null,
 * never 0. The curve is never extrapolated — YouTube reports it from 1% of the
 * video onwards, so the first scene's start retention is unknown.
 *
 * Pure (no network, no DB); never throws on malformed rows.
 */

import { sceneIdFor, type VideoScene } from "@/lib/storyboard";

/** Mirrors retention_analyzer.MIN_POINTS: fewer measured points describe nothing. */
export const MIN_POINTS = 5;
/** How many of the fastest-losing scenes the Storyboard highlights. */
export const HIGHLIGHT_WORST = 3;

const EPS = 1e-6;

/** One stored retention row (migration 0002); loose on purpose. */
export interface RetentionPointInput {
  elapsed_ratio?: number | null;
  watch_ratio?: number | null;
  measured_date?: string | null;
}

/** A scene window as the mapping needs it. */
export interface SceneWindow {
  id?: string | null;
  start_s?: number | null;
  end_s?: number | null;
}

export interface SceneRetention {
  sceneId: string;
  startS: number | null;
  endS: number | null;
  /** Share of viewers still watching at the scene's start / end. */
  retentionStart: number | null;
  retentionEnd: number | null;
  /** retentionStart − retentionEnd (share-of-audience points; negative when the curve rises). */
  drop: number | null;
  /** drop per minute of scene. */
  dropPerMin: number | null;
  /** 1 = fastest loss per minute; null when the rate is unknown. */
  rank: number | null;
}

export type SceneRetentionStatus =
  /** At least one scene has a measured drop. */
  | "ok"
  /** No usable retention curve for this video (none yet, or too few points). */
  | "no_curve"
  /** A curve exists, but no scene has real times or the video's length is unknown. */
  | "no_timing";

export interface SceneRetentionSummary {
  status: SceneRetentionStatus;
  scenes: SceneRetention[];
  /** Scene id → its row, for the Storyboard. */
  byId: Map<string, SceneRetention>;
}

function num(v: unknown): number | null {
  if (typeof v !== "number" || !Number.isFinite(v)) return null;
  return v;
}

function round4(v: number | null): number | null {
  return v === null ? null : Math.round(v * 1e4) / 1e4;
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/**
 * The newest measured curve as sorted [elapsed_ratio, watch_ratio] pairs, or []
 * when fewer than MIN_POINTS are usable. A null watch ratio is dropped, not read
 * as 0; only the newest measured_date counts; a repeated ratio keeps its last row.
 */
export function cleanCurve(points: readonly RetentionPointInput[] | null | undefined): Array<[number, number]> {
  let rows = (Array.isArray(points) ? points : []).filter(isObject) as RetentionPointInput[];
  const dates = rows.map((p) => p.measured_date).filter((d): d is string => typeof d === "string" && d !== "");
  if (dates.length) {
    const newest = dates.reduce((a, b) => (b > a ? b : a));
    rows = rows.filter((p) => !p.measured_date || p.measured_date === newest);
  }
  const byRatio = new Map<number, number>();
  for (const p of rows) {
    const r = num(p.elapsed_ratio);
    const w = num(p.watch_ratio);
    if (r === null || w === null || r < 0) continue;
    byRatio.set(r, w);
  }
  const curve = [...byRatio.entries()].sort((a, b) => a[0] - b[0]);
  return curve.length >= MIN_POINTS ? curve : [];
}

/** Retention at `ratio`, interpolated between measured points; null outside them. */
export function interpolate(curve: ReadonlyArray<[number, number]>, ratio: number | null): number | null {
  if (!curve.length || ratio === null) return null;
  const lo = curve[0][0];
  const hi = curve[curve.length - 1][0];
  if (ratio < lo - EPS || ratio > hi + EPS) return null;
  if (ratio <= lo) return curve[0][1];
  if (ratio >= hi) return curve[curve.length - 1][1];
  for (let i = 0; i + 1 < curve.length; i++) {
    const [r0, w0] = curve[i];
    const [r1, w1] = curve[i + 1];
    if (r0 <= ratio && ratio <= r1) {
      return r1 === r0 ? w1 : w0 + ((w1 - w0) * (ratio - r0)) / (r1 - r0);
    }
  }
  return null;
}

/**
 * The rendered video's length in seconds: the IR's measured audio duration,
 * else the last real scene end (the timeline the IR derives it from). Null when
 * neither exists — never a word-count estimate.
 */
export function videoDuration(manifest: unknown, scenes: readonly SceneWindow[] | null | undefined): number | null {
  const audio = isObject(manifest) ? manifest.audio : null;
  const d = isObject(audio) ? num(audio.duration_s) : null;
  if (d !== null && d > 0) return d;
  const ends = (Array.isArray(scenes) ? scenes : [])
    .map((s) => (isObject(s) ? num(s.end_s) : null))
    .filter((e): e is number => e !== null && e > 0);
  return ends.length ? Math.max(...ends) : null;
}

/** Rank by fastest loss per minute, then larger raw drop, then scene order. */
function rank(rows: SceneRetention[]): SceneRetention[] {
  const order = rows
    .map((r, i) => ({ r, i }))
    .filter(({ r }) => r.dropPerMin !== null)
    .sort((a, b) => b.r.dropPerMin! - a.r.dropPerMin! || (b.r.drop ?? 0) - (a.r.drop ?? 0) || a.i - b.i);
  const rankOf = new Map<number, number>();
  order.forEach(({ i }, k) => rankOf.set(i, k + 1));
  return rows.map((r, i) => ({ ...r, rank: rankOf.get(i) ?? null }));
}

/** Map a curve onto scene windows: one row per scene, null where unmeasurable. */
export function mapScenes(
  scenes: readonly SceneWindow[] | null | undefined,
  points: readonly RetentionPointInput[] | null | undefined,
  durationS: number | null,
): SceneRetention[] {
  const curve = cleanCurve(points);
  const duration = num(durationS) !== null && durationS! > 0 ? durationS! : null;
  const out: SceneRetention[] = [];
  (Array.isArray(scenes) ? scenes : []).forEach((s, position) => {
    if (!isObject(s)) return;
    const start = num(s.start_s);
    const end = num(s.end_s);
    let rStart: number | null = null;
    let rEnd: number | null = null;
    let drop: number | null = null;
    let rate: number | null = null;
    if (curve.length && duration !== null && start !== null && end !== null && end > start) {
      rStart = interpolate(curve, start / duration);
      rEnd = interpolate(curve, end / duration);
      if (rStart !== null && rEnd !== null) {
        drop = rStart - rEnd;
        rate = drop / ((end - start) / 60);
      }
    }
    out.push({
      sceneId: sceneIdFor(s as VideoScene, position),
      startS: start,
      endS: end,
      retentionStart: round4(rStart),
      retentionEnd: round4(rEnd),
      drop: round4(drop),
      dropPerMin: round4(rate),
      rank: null,
    });
  });
  return rank(out);
}

/**
 * Everything the video page needs: the per-scene rows keyed by the same scene
 * id the Storyboard uses, and why there is nothing to show when there isn't.
 */
export function summarizeSceneRetention(
  scenes: readonly VideoScene[] | null | undefined,
  manifest: unknown,
  points: readonly RetentionPointInput[] | null | undefined,
): SceneRetentionSummary {
  const list = Array.isArray(scenes) ? scenes : [];
  const rows = mapScenes(list, points, videoDuration(manifest, list));
  const status: SceneRetentionStatus = rows.some((r) => r.dropPerMin !== null)
    ? "ok"
    : cleanCurve(points).length === 0
      ? "no_curve"
      : "no_timing";
  return { status, scenes: rows, byId: new Map(rows.map((r) => [r.sceneId, r])) };
}

/** True when this scene is among the worst HIGHLIGHT_WORST AND actually loses viewers. */
export function isWorstScene(r: SceneRetention | null | undefined): boolean {
  return !!r && r.rank !== null && r.rank <= HIGHLIGHT_WORST && (r.drop ?? 0) > 0;
}

/** Bar width (0–100) of a scene's loss rate against the steepest scene's; null when unknown or not a loss. */
export function dropBarPercent(r: SceneRetention | null | undefined, all: readonly SceneRetention[]): number | null {
  if (!r || r.dropPerMin === null || r.dropPerMin <= 0) return null;
  const max = Math.max(...all.map((x) => x.dropPerMin ?? 0));
  return max > 0 ? Math.max(2, Math.round((r.dropPerMin / max) * 100)) : null;
}

/** A share (0.87) as a whole percent ("87%"); "—" when unknown. */
export function pctText(v: number | null): string {
  return v === null ? "—" : `${Math.round(v * 100)}%`;
}

/** A share-of-audience change in points with a sign ("−7.0", "+1.2"); "—" when unknown. */
export function pointsText(v: number | null): string {
  if (v === null) return "—";
  const p = v * 100;
  const sign = p > 0 ? "−" : p < 0 ? "+" : "";
  return `${sign}${Math.abs(p).toFixed(1)}`;
}

File: modules/scene_retention.py
"""Scene-level retention — which scene the audience left in.

Roadmap Q8 / PR 5.1. `retention_analyzer` says *where through the video*
viewers leave ("the biggest cliff starts around 40%"). A writer or a director
cannot act on a percentage; they act on a scene. The Video IR (migration 0013)
gives every scene its REAL ``start_s``/``end_s`` on the narration audio, and the
rendered video is exactly as long as that audio (the compositor sets the final
clip's duration to the audio's), so a point of the YouTube curve at
``elapsed_ratio`` sits at ``elapsed_ratio * duration`` seconds — inside exactly
one scene.

Per scene this yields:

* ``retention_start`` / ``retention_end`` — share of viewers still watching at
  the scene's first and last second, linearly interpolated between the two
  measured points around it;
* ``drop`` — ``start - end``, in share-of-audience points (negative when the
  curve rises, e.g. rewatches);
* ``drop_per_min`` — the same, per minute of scene. Worst scenes are RANKED on
  this, not on ``drop``: a long scene loses more viewers simply by being long,
  and a ranking on raw drop would only rediscover the longest scenes.

Honesty rules (CLAUDE.md #5)
----------------------------
* **Unknown stays unknown.** A scene without real times, a video whose duration
  is unknown, or a curve with fewer than ``retention_analyzer.MIN_POINTS``
  measured points yields ``None`` values — never 0.
* **Interpolate, never extrapolate.** YouTube reports the curve from 1% of the
  video onwards, so the first second of the first scene lies BEFORE the first
  measured point: its start retention is unknown, and so are its drop and rank.
  Filling that gap with "100% at 0 s" would be an assumption rendered as a
  measurement.
* Only the curve's newest ``measured_date`` is used (the same rule as
  ``StateStore.retention_curve``) — two snapshots of one curve are not two
  curves.

The TypeScript twin is ``command-center/lib/sceneRetention.ts``; both are
checked against the same cases in ``samples/scene_retention_cases.json``.

Pure and never raises: malformed input degrades to "unknown".
"""

from __future__ import annotations

import logging
import math
from dataclasses import dataclass
from typing import Iterable, List, Mapping, Optional, Sequence, Tuple

from modules.retention_analyzer import MIN_POINTS

logger = logging.getLogger(__name__)

#: A ratio this close outside the measured range still counts as inside it —
#: float rounding of ``end_s / duration`` must not turn the last scene's end
#: (exactly 100%) into "past the curve".
_EPS = 1e-6
#: Decimal places the public values are rounded to.
_DIGITS = 4


@dataclass(frozen=True)
class SceneRetention:
    scene_id: str
    index: int
    type: Optional[str]
    recipe: Optional[str]
    start_s: Optional[float]
    end_s: Optional[float]
    #: Share of viewers still watching at the scene's start / end (0.0-1.0+).
    retention_start: Optional[float]
    retention_end: Optional[float]
    #: retention_start - retention_end, share-of-audience points.
    drop: Optional[float]
    #: drop per minute of scene.
    drop_per_min: Optional[float]
    #: 1 = the scene that loses viewers fastest; None when drop_per_min is.
    rank: Optional[int] = None

    def to_dict(self) -> dict:
        return {
            "scene_id": self.scene_id, "index": self.index, "type": self.type,
            "recipe": self.recipe, "start_s": self.start_s, "end_s": self.end_s,
            "retention_start": self.retention_start, "retention_end": self.retention_end,
            "drop": self.drop, "drop_per_min": self.drop_per_min, "rank": self.rank,
        }


def _num(value) -> Optional[float]:
    if value is None or isinstance(value, bool):
        return None
    try:
        f = float(value)
    except (TypeError, ValueError):
        return None
    return f if math.isfinite(f) else None


def _str(value) -> Optional[str]:
    if value is None:
        return None
    s = str(value).strip()
    return s or None


def _round(value: Optional[float]) -> Optional[float]:
    return None if value is None else round(value, _DIGITS)


def clean_curve(points: Optional[Iterable[Mapping]]) -> List[Tuple[float, float]]:
    """The newest measured curve as sorted ``(elapsed_ratio, watch_ratio)``
    pairs, or ``[]`` when it has fewer than MIN_POINTS usable points.

    Rows without a measured ``watch_ratio`` are dropped (not read as 0). When
    rows span several ``measured_date`` values only the newest date is kept; a
    repeated ratio keeps its last row.
    """
    rows = [p for p in (points or []) if isinstance(p, Mapping)]
    dates = [str(p.get("measured_date")) for p in rows if p.get("measured_date")]
    if dates:
        newest = max(dates)
        rows = [p for p in rows if not p.get("measured_date") or str(p.get("measured_date")) == newest]
    by_ratio = {}
    for p in rows:
        ratio, watch = _num(p.get("elapsed_ratio")), _num(p.get("watch_ratio"))
        if ratio is None or watch is None or ratio < 0:
            continue
        by_ratio[ratio] = watch
    curve = sorted(by_ratio.items())
    return curve if len(curve) >= MIN_POINTS else []


def interpolate(curve: Sequence[Tuple[float, float]], ratio: Optional[float]) -> Optional[float]:
    """Retention at ``ratio``, linearly interpolated between the measured points
    around it; None outside the measured range (never extrapolated)."""
    if not curve or ratio is None:
        return None
    lo_r, hi_r = curve[0][0], curve[-1][0]
    if ratio < lo_r - _EPS or ratio > hi_r + _EPS:
        return None
    if ratio <= lo_r:
        return curve[0][1]
    if ratio >= hi_r:
        return curve[-1][1]
    for (r0, w0), (r1, w1) in zip(curve, curve[1:]):
        if r0 <= ratio <= r1:
            if r1 == r0:
                return w1
            return w0 + (w1 - w0) * (ratio - r0) / (r1 - r0)
    return None  # unreachable for a sorted curve; unknown rather than a guess


def video_duration(manifest: Optional[Mapping], scenes: Optional[Sequence[Mapping]] = None) -> Optional[float]:
    """The rendered video's length in seconds, or None when unknown.

    The IR's measured audio duration first; otherwise the last scene's real end
    — the same timeline the IR itself derives the audio duration from when the
    mixer did not report one. Never a word-count estimate.
    """
    try:
        audio = (manifest or {}).get("audio") if isinstance(manifest, Mapping) else None
        d = _num(audio.get("duration_s")) if isinstance(audio, Mapping) else None
        if d is not None and d > 0:
            return d
        ends = [_num(s.get("end_s")) for s in (scenes or []) if isinstance(s, Mapping)]
        ends = [e for e in ends if e is not None and e > 0]
        return max(ends) if ends else None
    except Exception:
        return None


def _scene_id(scene: Mapping, position: int) -> str:
    sid = _str(scene.get("id"))
    return sid or f"s{position:03d}"


def _recipe(scene: Mapping) -> Optional[str]:
    shot = scene.get("shot")
    return _str(shot.get("recipe")) if isinstance(shot, Mapping) else None


def rank_scenes(rows: Sequence[SceneRetention]) -> List[SceneRetention]:
    """``rows`` with ``rank`` set: 1 = fastest loss per minute, then larger raw
    drop, then earlier scene. Scenes with an unknown rate stay unranked."""
    known = [r for r in rows if r.drop_per_min is not None]
    order = sorted(known, key=lambda r: (-r.drop_per_min, -(r.drop or 0.0), r.index))
    rank_of = {id(r): i + 1 for i, r in enumerate(order)}
    return [
        SceneRetention(**{**r.__dict__, "rank": rank_of.get(id(r))}) for r in rows
    ]


def map_scenes(
    scenes: Optional[Sequence[Mapping]],
    points: Optional[Iterable[Mapping]],
    duration_s: Optional[float],
) -> List[SceneRetention]:
    """Map a retention curve onto scene windows. One entry per scene, in order;
    every value None where it cannot be measured. Never raises."""
    try:
        curve = clean_curve(points)
        duration = _num(duration_s)
        if duration is not None and duration <= 0:
            duration = None
        out = []
        for position, scene in enumerate(scenes or []):
            if not isinstance(scene, Mapping):
                continue
            start, end = _num(scene.get("start_s")), _num(scene.get("end_s"))
            r_start = r_end = drop = rate = None
            if curve and duration is not None and start is not None and end is not None and end > start:
                r_start = interpolate(curve, start / duration)
                r_end = interpolate(curve, end / duration)
                if r_start is not None and r_end is not None:
                    drop = r_start - r_end
                    rate = drop / ((end - start) / 60.0)
            idx = scene.get("index")
            out.append(SceneRetention(
                scene_id=_scene_id(scene, position),
                index=int(idx) if isinstance(idx, int) and not isinstance(idx, bool) else position,
                type=_str(scene.get("type")), recipe=_recipe(scene),
                start_s=start, end_s=end,
                retention_start=_round(r_start), retention_end=_round(r_end),
                drop=_round(drop), drop_per_min=_round(rate),
            ))
        return rank_scenes(out)
    except Exception:
        logger.warning("scene_retention: mapping failed; every scene reads as unknown", exc_info=True)
        return []


def for_video(manifest: Optional[Mapping], points: Optional[Iterable[Mapping]]) -> List[SceneRetention]:
    """Scene retention for one video from its stored Video IR (``videos.manifest``)
    and its retention rows. [] when there is no manifest."""
    if not isinstance(manifest, Mapping):
        return []
    scenes = [s for s in (manifest.get("scenes") or []) if isinstance(s, Mapping)]
    return map_scenes(scenes, points, video_duration(manifest, scenes))


def has_data(rows: Sequence[SceneRetention]) -> bool:
    """True when at least one scene has a measured drop rate."""
    return any(r.drop_per_min is not None for r in rows or [])


def median(values: Sequence[float]) -> Optional[float]:
    vals = sorted(v for v in values if v is not None)
    if not vals:
        return None
    mid = len(vals) // 2
    return vals[mid] if len(vals) % 2 else (vals[mid - 1] + vals[mid]) / 2
